stable_baselines3/her/her.py

```diff
@@ -0,0 +1,495 @@
+from stable_baselines3.common.policies import BasePolicy
+from stable_baselines3.common.save_util import load_from_zip_file, recursive_getattr, recursive_setattr
+from stable_baselines3.common.type_aliases import GymEnv, MaybeCallback, RolloutReturn
+from stable_baselines3.common.utils import check_for_correct_spaces
+from stable_baselines3.common.vec_env import VecEnv
+from stable_baselines3.common.vec_env.obs_dict_wrapper import ObsDictWrapper
+from stable_baselines3.her.goal_selection_strategy import KEY_TO_GOAL_STRATEGY, GoalSelectionStrategy
+from stable_baselines3.her.her_replay_buffer import HerReplayBuffer
+
+
+def get_time_limit(env: VecEnv, current_max_episode_length: Optional[int]) -> int:
+    """
+    Get time limit from environment.
+
+    :param env: Environment from which we want to get the time limit.
+    :param current_max_episode_length: Current value for max_episode_length.
+    :return: max episode length
+    """
+    # try to get the attribute from environment
+    if current_max_episode_length is None:
+        try:
+            current_max_episode_length = env.get_attr("spec")[0].max_episode_steps
+        # if not available check if a valid value was passed as an argument
+        except AttributeError:
+            raise ValueError(
+                "The max episode length could not be inferred."
+                "You must specify a `max_episode_steps` when registering the environment, "
+                "use a `gym.wrappers.TimeLimit` wrapper "
+                "or pass `max_episode_length` to the model constructor"
+            )
+    return current_max_episode_length
+
+
+class HER(BaseAlgorithm):
+    """
+    Hindsight Experience Replay (HER)
+
+    Paper: https://arxiv.org/abs/1707.01495
+
+    :param policy: The policy model to use.
+    :param env: The environment to learn from (if registered in Gym, can be str)
+    :param model_class: Off policy model which will be used with hindsight experience replay. (SAC, TD3)
+    :param n_sampled_goal: Number of sampled goals for replay. (offline sampling)
+    :param goal_selection_strategy: Strategy for sampling goals for replay.
+        One of ['episode', 'final', 'future', 'random']
+    :param online_sampling: Sample HER transitions online.
+    :param learning_rate: learning rate for the optimizer,
+        it can be a function of the current progress remaining (from 1 to 0)
+    :param max_episode_length: The maximum length of an episode. If not specified,
+        it will be automatically inferred if the environment uses a ``gym.wrappers.TimeLimit`` wrapper
+    """
+
+    def __init__(
+        self,
+        policy: Union[str, Type[BasePolicy]],
+        env: Union[GymEnv, str],
+        model_class: Type[OffPolicyAlgorithm],
+        n_sampled_goal: int = 5,
+        goal_selection_strategy: Union[GoalSelectionStrategy, str] = "future",
+        online_sampling: bool = False,
+        learning_rate: Union[float, Callable] = 3e-4,
```

Comment:
I know this is required by the base class, but we should use the default from the offline model

so probably the best way to achieve that is to pass a dummy value to the base class and override it once the model is initialized